reader.go

```diff
@@ -168,6 +175,10 @@ func (r *Reader) ReadMessage(ctx context.Context) (msg Message, err error) {
 			err = ctx.Err()
 			return
 
+		case watermark := <-r.highWatermarks:
```

Comment:
I'm not a big fan of this, it seems like we could pass it as part of the internal `readerMessage` data structure.